flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/lookup/FileStoreLookupFunction.java

```diff
@@ -0,0 +1,216 @@
+
+        this.projectFields =
+                Arrays.stream(projection)
+                        .mapToObj(i -> schema.fieldNames().get(i))
+                        .collect(Collectors.toList());
+
+        // add primary keys
+        for (String field : schema.primaryKeys()) {
+            if (!projectFields.contains(field)) {
+                projectFields.add(field);
+            }
+        }
+
+        this.predicate = predicate;
+    }
+
+    @Override
+    public void open(FunctionContext context) throws Exception {
+        super.open(context);
+        String tmpDirectory = getTmpDirectory(context);
+        this.path = new File(tmpDirectory, "lookup-" + UUID.randomUUID());
+
+        Configuration options = Configuration.fromMap(table.schema().options());
+        this.refreshInterval = options.get(CoreOptions.CONTINUOUS_DISCOVERY_INTERVAL);
+        this.stateFactory = new RocksDBStateFactory(path.toString(), options);
+
+        List<String> fieldNames = table.schema().logicalRowType().getFieldNames();
+        int[] projection = projectFields.stream().mapToInt(fieldNames::indexOf).toArray();
+        RowType rowType = TypeUtils.project(table.schema().logicalRowType(), projection);
+
+        PredicateFilter recordFilter = createRecordFilter(projection);
+        this.lookupTable =
+                LookupTable.create(
+                        stateFactory,
+                        rowType,
+                        table.schema().primaryKeys(),
+                        joinKeys,
+                        recordFilter,
+                        options.getLong(LRU_CACHE_MAX_ROWS));
+        this.nextLoadTime = -1;
+        this.streamingReader = new TableStreamingReader(table, projection, this.predicate);
+
+        // do first load
+        refresh();
+    }
+
+    private PredicateFilter createRecordFilter(int[] projection) {
+        Predicate adjustedPredicate = null;
+        if (predicate != null) {
+            // adjust to projection index
+            adjustedPredicate =
+                    transformFieldMapping(
+                                    this.predicate,
+                                    IntStream.range(0, table.schema().fields().size())
+                                            .map(i -> Ints.indexOf(projection, i))
+                                            .toArray())
+                            .orElse(null);
+        }
+        return new PredicateFilter(
+                TypeUtils.project(table.schema().logicalRowType(), projection), adjustedPredicate);
+    }
+
+    /** Used by code generation. */
+    @SuppressWarnings("unused")
+    public void eval(Object... values) throws IOException {
+        checkRefresh();
+        List<RowData> results = lookupTable.get(GenericRowData.of(values));
+        for (RowData matchedRow : results) {
+            collect(matchedRow);
+        }
+    }
+
+    private void checkRefresh() throws IOException {
+        if (nextLoadTime > System.currentTimeMillis()) {
```

Comment:
Does this check only log that lookup table has refreshed?

Reply:
yes